solr/core/src/java/org/apache/solr/handler/admin/ClusterStatus.java

```diff
@@ -205,54 +214,32 @@ private void fetchClusterStatusForCollOrAlias(
       }
     }
 
-    // TODO use an Iterable to stream the data to the client instead of gathering it all in mem
-
-    NamedList<Object> collectionProps = new SimpleOrderedMap<>();
-
-    collectionStream.forEach(
-        clusterStateCollection -> {
-          Map<String, Object> collectionStatus;
-          String name = clusterStateCollection.getName();
-
-          Set<String> requestedShards = new HashSet<>();
-          if (routeKey != null) {
-            DocRouter router = clusterStateCollection.getRouter();
-            Collection<Slice> slices =
-                router.getSearchSlices(routeKey, null, clusterStateCollection);
-            for (Slice slice : slices) {
-              requestedShards.add(slice.getName());
-            }
-          }
-          if (shard != null) {
-            String[] paramShards = shard.split(",");
-            requestedShards.addAll(Arrays.asList(paramShards));
-          }
-
-          byte[] bytes = Utils.toJSON(clusterStateCollection);
-          @SuppressWarnings("unchecked")
-          Map<String, Object> docCollection = (Map<String, Object>) Utils.fromJSON(bytes);
-          collectionStatus = getCollectionStatus(docCollection, name, requestedShards);
-
-          collectionStatus.put("znodeVersion", clusterStateCollection.getZNodeVersion());
-          collectionStatus.put(
-              "creationTimeMillis", clusterStateCollection.getCreationTime().toEpochMilli());
-
-          if (collectionVsAliases.containsKey(name) && !collectionVsAliases.get(name).isEmpty()) {
-            collectionStatus.put("aliases", collectionVsAliases.get(name));
+    MapWriter collectionPropsWriter =
+        ew -> {
+          Iterator<Map<String, Object>> it =
+              collectionStream
+                  .map(
+                      (collectionState) ->
+                          collectionPropsResponse(
+                              collectionState,
+                              collectionVsAliases,
+                              routeKey,
+                              liveNodes,
+                              requestedShards))
+                  .iterator();
+          while (it.hasNext()) {
```

Comment:
Oh; now that I see you are iterating it yourself (I had thought that would be driven externally), you shouldn't bother calling .iterator then looping; you can loop directly with stream.forEach